pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,244 @@
+
+As an example of the relationship between schema resources and lexical scopes,
+consider the `$anchor` keyword that defines a [location-independent
+identifier](https://json-schema.org/draft/2020-12/json-schema-core#section-8.2.2)
+for a schema. While declaring the same anchor identifier more than once in the
+same schema resource is an error, it is possible to declare the same anchor
+identifier on different schema resources within the same schema. This is
+possible because <i>schema resources do not share their lexical scopes</i>:
+
+![Defining multiple anchors with the same name on different schema resources](/img/posts/2024/understanding-lexical-dynamic-scopes/static-anchors.webp)
+
+### Changing Scopes With References
+
+Now that we understand lexical scopes and their boundaries within schema
+resources, let's consider the effect that following references has on them.
+<i>When evaluation encounters a reference keyword, it abandons the lexical scope
+of the reference schema and enters the lexical scope of the destination
+schema</i>. In other words, the evaluation process needs to re-calculate the
+lexical scope of the schema it ends up at after following the reference.
+
+Consider the following example. In the left, the lexical scope of a subschema
+that defines a static reference. In the right, the lexical scope of the
+destination subschema after following such reference. Note that the destination
+is part of a different schema resource, under a different meta-schema. The
+change of lexical scope let's the destination schema be governed, as expected,
+by its own meta-schema instead of by the meta-schema of the root schema
+resource:
+
+![Lexical scope before and after following a static reference](/img/posts/2024/understanding-lexical-dynamic-scopes/ref.webp)
+
+As an interesting scenario, note that the root schema resource is not
+necessarily in the lexical scope of the first subschema that the evaluation
+process will visit. Consider a remote reference that points to a nested schema
+resource. In this case, the evaluation process jumps to the lexical scope of
+the destination schema resource, which may not include the root schema
+resource. For example:
+
+![Following a remote reference to a nested schema resource](/img/posts/2024/understanding-lexical-dynamic-scopes/ref-anchor.webp)
+
+Dynamic Scope
+-------------
+
+To recap, the lexical scope of a schema consists of the schema itself and its
+corresponding parent schemas within the same schema resource. In comparison,
+<i>the dynamic scope of a schema consists of the subschemas that were evaluated
+so far, including the schema itself, regardless of the schema resource they
+belong to</i>.
+
+Coming back to our analogy of a JSON Schema as a tree structure, the dynamic
+scope corresponds to the sequence of nodes that were visited by the evaluation
+process. In JSON Schema parlance, this is referred to as the <i>evaluation path</i>.
```

Comment:
I don't think "resources" is correct when discussing the general concept of dynamic scope rather than the implementation of dynamic references where it's more practical to track resources.

The dynamic scope is the stack of subschema locations visited while evaluating the schema. Anchors/dynamicAnchor a weird class of keywords that apply whenever you enter a resource (also includes `$id` and `$schema`). So, no matter what subschema I enter the resource through, any anchors in that resource are going to be in scope. That's why it's only necessary to track resources to implement dynamic references even though the dynamic scope is actually the visited subschemas.

Reply:
(I'll add that I hate that dynamic scope is defined this way. It's unhelpful and unnecessary. It only serves to make dynamic references harder to understand. It should be a resource-level concept IMO.)

Reply:
> The dynamic scope is the stack of subschema locations visited while evaluating the schema. 

So if I'm following correctly, is the definition I'm providing in the article technically correct (even if we want to explain it in some other way)? By what @jdesrosiers is saying, it is indeed the stack of visited subschemas, not the stack of schema resources

Reply:
> That's why it's only necessary to track resources to implement dynamic references even though the dynamic scope is actually the visited subschemas. @jdesrosiers 

What's the benefit of defining a dynamic resource in a way that doesn't help you resolve dynamic references?  Supporting dynamic references is _why_ we defined dynamic scope.

> So, no matter what subschema I enter the resource through, any anchors in that resource are going to be in scope.

If the anchors _anywhere in the resource_ are in scope, then that tells me that the entire resource must be part of the scope.  Thus dynamic scope is the ordered set of resources identified by the evaluation path.

Reply:
@gregsdennis, I agree with everything you said. It doesn't make a lot of sense. You'll have to ask Henry why he defined it that way.